liquibase/liquibase
# Flag to Control Logging of SQL Statements and SQL Execution Output

Adds a `liquibase.sql.logLevel` configuration option to control the level at which sql is logged. The valueProvider logic allows it to be set through the CLI, env variables, etc.

It defaults to "FINE" like the 4.x code currently does, but can also be set to "INFO" to act like it did in 3.x or even SEVERE if needed.

An additional feature this provides is the ability to set it to OFF which will block all SQL logging which can be useful for people who do not want their sql being logged in a shared environment.

Provides as work-around for #1501

## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: Stand alone PR
* Local Branch: [https://github.com/liquibase/liquibase/tree/DAT-5863](https://github.com/liquibase/liquibase/tree/DAT-5863)
- Test Results: https://github.com/liquibase/liquibase/pull/1520/checks

#### Testing
* Steps to Reproduce: See above
* Guidance:
  * Does not add more to what would get logged, just controls what level existing SQL messages are be logged at
  * Check is in that does the logging, not in the logger. So should apply to any logger implementation used
  * Any of our log levels can be set: FINE, INFO, SEVERE, WARNING, OFF
  * Example using CLI arguments: `liquibase --sql-log-level=WARNING --log-level=WARNING update` 
  * We don't force the --log-level flag to change based on the --sql-log-level flag, we still respect what that is set to (or not set)
  * Example log message: `[2021-10-22 11:13:52] WARNING [liquibase.executor] CREATE TABLE LBUSER.test_table (id INTEGER, test_col CLOB)` when --sql-log-level=WARNING

## Test Requirements (Internal Liquibase QA)
The change this fix introduces is to allow the user to set the log level that SQL generated by Liquibase is output to the log. Prior to this fix, SQL was output at the same setting as the global --log-level setting. The contents of the output remains the same; it is only the log level that is different.

The log messages impacted are ones executed by `liquibase.executor`. An example of a log message generated by Liquibase is:

`SEVERE [liquibase.executor] CREATE FUNCTION [dbo].[emailFunction]()`

The SQL log messages generated by databases are not impacted. An example of a SQL log message that is generated by a database is:

`FINE [com.microsoft.sqlserver.jdbc.SQLServerStatement doExecuteStatement] SQLServerStatement:32 Executing (not server cursor) CREATE FUNCTIon[dbo].[emailFunction]()`

This fix does not depend on the database platform.

Any valid changelog can be used for these tests; make sure that at least one changeset deploys with each update in order to have SQL statements that can be written to the log.

**Manual Tests**

_Verify liquibase update outputs liquibase.executor messages at the expected log level._

`liquibase --sql-log-level <level> --log-level <level> --log-file log.txt update --changelog-file <changelog>`

For each test case, substitute <level> for sql-log-level and log-level with the values in the table.

|**sql~~log~~level**|**log-level**|**Expected**|**P/F**|
|----|----|----|----|
|FINE|INFO|FINE| |
|INFO|INFO|INFO| |
|WARNING|INFO|WARNING| |
|SEVERE|INFO|SEVERE| |
|DEBUG|INFO|FINE| |

|OFF|FINE|No SQL logging for lb.executor
Informational/non-SQL logging for lb.executor| |


|----|----|----|----|
|FOO|INFO|Unknown log level FOO| |

_Verify update with sql~~log~~level configured sends logs to Hub, with SQL at the expected log level._ 

**Setup**

* Create an account on Hub Staging.
  * Go to the Liquibase Hub Staging page 
  * Sign up for an account.
  * Check your email and follow the link to activate an account.
* Register a changelog with hub-staging.
  * `liquibase register-changelog --changelog-file <file>`
  * When prompted, create a new project named LB884
* Edit the liquibase.properties file to add the following properties:

```
# HUB PROPERTIES #
liquibase.hub.url: <Hub Staging URL>
liquibase.hub.apiKey: <YOUR_STAGING_API>
liquibase.hub.logLevel: INFO
liquibase.hub.mode: ALL
```

Now, execute the test case:
`liquibase --sql-log-level SEVERE --log-level INFO --log-file log.txt update --changelog-file <changelog>`

* The update is successful.
* A log is sent to Hub
  * Find this on the Projects/LB884/Operations by clicking on the little ‘download’ icon at the bottom right of the operation row corresponding to the update just executed.
* The log has liquibase.executor SQL statements logged at SEVERE.
* All other log output is logged at INFO.

**Automated Tests**

* No new functional tests are required for this ticket.
* No new test-harness tests are required for this ticket.
* New unit tests were added here: liquibase-core/src/test/groovy/liquibase/configuration/ConfigurationValueConverterTest.groovy

#### Dev Verification
(did not save output at the time)



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-884) by [Unito](https://www.unito.io)
